builtin/logical/database/dbplugin/databasemiddleware.go

```diff
@@ -0,0 +1,172 @@
+package dbplugin
+
+import (
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+	log "github.com/mgutz/logxi/v1"
+)
+
+// ---- Tracing Middleware Domain ----
+
+// databaseTracingMiddleware wraps a implementation of DatabaseType and executes
+// trace logging on function call.
+type databaseTracingMiddleware struct {
+	next   DatabaseType
+	logger log.Logger
+
+	typeStr string
+}
+
+func (mw *databaseTracingMiddleware) Type() (string, error) {
+	return mw.next.Type()
+}
+
+func (mw *databaseTracingMiddleware) CreateUser(statements Statements, usernamePrefix string, expiration time.Time) (username string, password string, err error) {
+	if mw.logger.IsTrace() {
+		defer func(then time.Time) {
+			mw.logger.Trace("database/CreateUser: finished", "type", mw.typeStr, "err", err, "took", time.Since(then))
+		}(time.Now())
+
+		mw.logger.Trace("database/CreateUser: starting", "type", mw.typeStr)
+	}
+	return mw.next.CreateUser(statements, usernamePrefix, expiration)
+}
+
+func (mw *databaseTracingMiddleware) RenewUser(statements Statements, username string, expiration time.Time) (err error) {
+	if mw.logger.IsTrace() {
+		defer func(then time.Time) {
+			mw.logger.Trace("database/RenewUser: finished", "type", mw.typeStr, "err", err, "took", time.Since(then))
```

Comment:
I've gotten some heat in the past for this type of logging (`database/RenewUser`), which mostly got peppered into code *before* we moved to more structured logging. I've been removing that here and there over time. Let's switch this up so that it's `database` and put in a parameter that specifies the operation type, e.g. `...., "operation", "renew_user", ....` or so.